### PR TITLE
Fix package install instructions in "Numerical Sage" tutorial

### LINKED ISSUE #19198: Installation instructions in numerical_sage tutorial are obsolete
<div id="comment:0"></div>

The following files

* [src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst](http://doc.sagemath.org/html/en/thematic_tutorials/numerical_sage/mpi4py.html)
* [src/doc/en/thematic_tutorials/numerical_sage/installation_linux.rst](http://doc.sagemath.org/html/en/thematic_tutorials/numerical_sage/installation_linux.html)
* [src/doc/en/thematic_tutorials/numerical_sage/installation_osx.rst](http://doc.sagemath.org/html/en/thematic_tutorials/numerical_sage/installation_osx.html)

contain instructions to install a laundry list of packages.

However, those instructions don't actually work, for 2 reasons:
1. they use `install_package()` which no longer works (#16760 and #16759)
2. the packages are no longer installable anyway (#19004)

CC:  @jhpalmieri

Component: **documentation**

_Issue created by migration from https://trac.sagemath.org/ticket/19198_

### PR DESCRIPTION
**Pull request by mkoeppe**
Fixes #19198

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 1]
```diff
@@ -59,18 +42,19 @@ example. Create a script with the following code in a file mpi_1.py
     print("hello world")
     print("my rank is: %d"%comm.rank)
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]: It is a good idea to update the syntax, i.e.
```suggestion
    print(f"my rank is: {comm.rank}")
```

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 2]
```diff
@@ -59,18 +42,19 @@ example. Create a script with the following code in a file mpi_1.py
...
+To run it you can do (from the command line in your Sage
 directory)
 
 .. CODE-BLOCK:: shell-session
 
-    ./local/bin/mpirun -np 5 ./sage -python mpi_1.py
+    mpirun -np 5 ./sage -python mpi_1.py
+
+The command ``mpirun -np 5`` starts 5 copies of a program under MPI. In
+this case we have 5 copies of Sage in pure Python mode run the
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]:
```suggestion
this case we have 5 copies of Sage in pure Python mode running the
```

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 3]
```diff
@@ -86,20 +70,20 @@ Consider the following example which you should put in a script mpi_2.py
     print("I received this:")
     print(data)
 
-The same command as above with mpi_1.py replaced by mpi_2.py will
+The same command as above with ``mpi_1.py`` replaced by ``mpi_2.py`` will
 produce 5 outputs and you will see each process creates an array and
 then passes it to the next guy (where the last guy passes to the
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]:
```suggestion
produce 5 outputs. Each process will create an array and pass
it to the next process, where the last process passes to the
```

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 4]
```diff
@@ -86,20 +70,20 @@ Consider the following example which you should put in a script mpi_2.py
...
-first.) Note that MPI.size is the total number of mpi
-processes. MPI.COMM WORLD is the communication world.
+first.) Note that ``MPI.size`` is the total number of MPI
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]:
```suggestion
first. Note that ``MPI.size`` is the total number of MPI
```

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 5]
```diff
@@ -126,8 +110,8 @@ ready to receive and then he will send and process 2 will be
 waiting to receive, etc. This will not lock regardless of how large
 of an array we pass.
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]: I can't seem to comment on L108-109, but replace `he` -> `it`. My suggestion:

```diff
- Now the first process initiates a send, and then process 1 will be
- ready to receive and then he will send and process 2 will be
- waiting to receive, etc. This will not lock regardless of how large
- of an array we pass.
+ Now, process 0 sends the data to process 1, then waits to receive from process ``MPI.size - 1``.
+ Simultaneously, process 1 will send the data to process 2, then receives the data from process 0.
+ This will not lock even if the array transmitted is huge.
```

(Please line wrap for me)

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 6]
```diff
@@ -126,8 +110,8 @@ ready to receive and then he will send and process 2 will be
...
 
-A common idiom is to have one process, usually the one with rank 0
-act as a leader. That processes sends data out to the other
+A common idiom is to have one process, usually the one with rank 0,
+act as a leader. That process sends data out to the other
 processes and processes the results and decides how further
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]:
```suggestion
processes, compute on the results, and decides how much further
```

File: src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst [hunk 7]
```diff
@@ -148,18 +132,18 @@ computation should proceed. Consider the following code
...
+everyone, but when root executes it, it acts as a ``send`` and a
+``receive`` (root gets one row from itself), while for everyone else it
+is just a ``receive``.
+
+There is a complementary ``gather`` command that collects results from
+all the processes into a list. The next example uses ``scatter`` and
+``gather`` together. Now the root process scatters the rows of a
 matrix, each process then squares the elements of the row it gets.
 Then the rows are all gathered up again by the root process who
 collects them into a new matrix.
```
Comment (grhkm21) [inline src/doc/en/thematic_tutorials/numerical_sage/mpi4py.rst]:
```suggestion
``gather`` together. Now the root process scatters the rows of a
matrix. Each process squares the elements of the row it receives.
The root process then gathers the rows into a new matrix.
```

**Review by grhkm21 — COMMENTED** [review]
Just grammar edits, hope you don't mind. It doesn't seem to be the focus of the PR, but might as well?

**Comment by mkoeppe** [comment]
Thanks a lot for the edits and suggestions. I've made more changes along these lines

**Comment by grhkm21** [comment]
Thanks, the new changes are great, I was worried I would get too annoying with my review changes.

Looks good to me.

P.S.

```diff
- MPI programming is difficult. It is "schizophrenic programming" in
- that you are writing a single programming with multiple threads of
- execution "many voices in one head".
```

This is quite funny.

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**